rust/module/src/lib.rs

```diff
@@ -0,0 +1,355 @@
+        string = string,
+    )
+}
+
+fn build_modinfo_string(module_name: &str, name: &str, string: &str) -> String {
+    format!(
+        "
+            // Loadable modules' modinfo strings go as-is
+            #[cfg(MODULE)]
+            #[link_section = \".modinfo\"]
+            #[used]
+            pub static {name}: [u8; {length}] = *b\"{string}\\0\";
+        ",
+        name = name,
+        length = string.len() + 1,
+        string = string,
+    ) + &build_modinfo_builtin_string(module_name, name, string)
+}
+
+/// Declares a kernel module.
+///
+/// The `typename` argument should be a type which implements the [`KernelModule`] trait.
+/// Also accepts various forms of kernel metadata.
+///
+/// Example:
+/// ```rust,no_run
+/// use kernel::prelude::*;
+///
+/// module!{
+///     typename: MyKernelModule,
+///     name: b"my_kernel_module",
+///     author: b"Rust for Linux Contributors",
+///     description: b"My very own kernel module!",
+///     license: b"GPL v2",
+///     params: {},
+/// }
+///
+/// struct MyKernelModule;
+///
+/// impl KernelModule for MyKernelModule {
+///     fn init() -> KernelResult<Self> {
+///         Ok(MyKernelModule)
+///     }
+/// }
+/// ```
+#[proc_macro]
+pub fn module(ts: TokenStream) -> TokenStream {
+    let mut it = ts.into_iter();
+
+    let typename = get_ident(&mut it, "typename");
+    let name = get_byte_string(&mut it, "name");
+    let author = get_byte_string(&mut it, "author");
+    let description = get_byte_string(&mut it, "description");
+    let license = get_byte_string(&mut it, "license");
+    let params = get_group(&mut it, "params");
+
+    expect_end(&mut it);
+
+    assert_eq!(params.delimiter(), Delimiter::Brace);
+
+    let mut it = params.stream().into_iter();
+
+    let mut params_modinfo = String::new();
+
+    loop {
+        let param_name = match it.next() {
+            Some(TokenTree::Ident(ident)) => ident.to_string(),
+            Some(_) => panic!("Expected Ident or end"),
+            None => break,
+        };
+
+        assert_eq!(expect_punct(&mut it), ':');
+        let param_type = expect_ident(&mut it);
+        let group = expect_group(&mut it);
+        assert_eq!(expect_punct(&mut it), ',');
+
+        assert_eq!(group.delimiter(), Delimiter::Brace);
+
+        let mut param_it = group.stream().into_iter();
+        let param_default = if param_type == "bool" {
+            get_ident(&mut param_it, "default_")
+        } else {
+            get_literal(&mut param_it, "default_")
+        };
+        let param_permissions = get_literal(&mut param_it, "permissions");
+        let param_description = get_byte_string(&mut param_it, "description");
+        expect_end(&mut param_it);
+
+        // TODO: more primitive types
+        // TODO: other kinds: arrays, unsafes, etc.
+        let param_kernel_type = match param_type.as_ref() {
+            "bool" => "bool",
+            "i32" => "int",
+            _ => panic!("Unrecognized type"),
```

Comment:
Can you include the type in the message.

Reply:
Sure, will do.